commit 30548c1433a4c33e6bbc64ff36f9b193a2942d54 (refs/changes/95/2195/3)
Author: Richard Bradley <richard.bradley@joyent.com>
Date:   2017-10-06T08:07:23+00:00 (2 years ago)
    
    MANATEE-341 "manatee-adm rebuild" hangs when rebuilding a sync peer
    MANATEE-344 Rebuild should not exit if postgres hasn't started on its first attempt

diff --git a/lib/adm.js b/lib/adm.js
index dd822ef..6e642b0 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -1507,11 +1507,19 @@ function rebuild(opts, cb) {
                     _addPostgresStatus,
                     function _getCurrentPeer(subctx, subcb) {
                         var thisPeer = null;
-                        subctx.state[_.shard].async.forEach(function (peer) {
-                            if (peer.zoneId == _.config.zoneId) {
-                                thisPeer = peer;
-                            }
-                        });
+
+                        if (subctx.state[_.shard].sync &&
+                            subctx.state[_.shard].sync.zoneId ===
+                            _.config.zoneId) {
+                            thisPeer = subctx.state[_.shard].sync;
+                        } else if (subctx.state[_.shard].async) {
+                            subctx.state[_.shard].async.forEach(function (p) {
+                                if (p.zoneId === _.config.zoneId) {
+                                    thisPeer = p;
+                                }
+                            });
+                        }
+
                         _.peer = thisPeer;
                         subcb();
                     },
@@ -1544,7 +1552,7 @@ function rebuild(opts, cb) {
                             process.stderr.write('\n');
                             if (restoreTry >= RESTORE_RETRIES) {
                                 return _cb(new Error('This Manatee instance ' +
-                                    'is not an active async after ' +
+                                    'is not an active peer after ' +
                                     RESTORE_RETRIES + ' restore attempts.  ' +
                                     'Check sitter logs.'));
                             }
@@ -1589,18 +1597,6 @@ function rebuild(opts, cb) {
                         }
 
                         lastRestore = restore;
-
-                    } else if (waitCount > 60) {
-                        /*
-                         * At this point no restore has taken place and
-                         * postgres hasn't started after roughly 60 seconds.
-                         * Something is likely wrong with postgres, so we error
-                         * here to prevent an unlimited spinner.
-                         */
-                        process.stderr.write('\n');
-                        return _cb(new Error('This Manatee instance is not ' +
-                            'active async after 60 seconds. Check sitter ' +
-                            'logs.'));
                     }
 
                     /*
